### PR TITLE
Positional Audio support for Battlefield 4 (x64)

### PR DESCRIPTION
**Pull request by davidebeatrici**
Context based on Server ID.
Identity data: IP address + port, team, squad and squad leader.

### CODE REVIEW

File: plugins/bf4/bf4.cpp [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+            oidentity << std::endl << "\"Squad\": \"X-ray\"";
+        else if (squad == 25)
+            oidentity << std::endl << "\"Squad\": \"Yankee\"";
+        else if (squad == 26)
+            oidentity << std::endl << "\"Squad\": \"Zulu\"";
+
+        if (squad_leader)
+            oidentity << std::endl << "\"Squad leader\": \"Yes\""; // If squad leader value is 1, set squad leader state to "Yes" as squad leader in identity
+        else
+            oidentity << std::endl << "\"Squad leader\": \"No\""; // If squad leader value is 1, set squad leader state to "No" as squad leader in identity
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Do you mean if the value is 0?

File: plugins/bf4/bf4.cpp [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            oidentity << std::endl << "\"Squad\": \"Whiskey\"";
+        else if (squad == 24)
+            oidentity << std::endl << "\"Squad\": \"X-ray\"";
+        else if (squad == 25)
+            oidentity << std::endl << "\"Squad\": \"Yankee\"";
+        else if (squad == 26)
+            oidentity << std::endl << "\"Squad\": \"Zulu\"";
+
+        if (squad_leader)
+            oidentity << std::endl << "\"Squad leader\": \"Yes\""; // If squad leader value is 1, set squad leader state to "Yes" as squad leader in identity
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Should this be a boolean instead of a string?

So

`"Yes" -> true`

and

`"No" -> false`?

File: plugins/bf4/bf4.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        for (int i=0;i<3;i++) {
+            avatar_pos[i] = avatar_front[i] = avatar_top[i] = camera_pos[i] =  camera_front[i] = camera_top[i] = 0.0f;
+        }
+
+        return true; // This tells Mumble to ignore all vectors.
+    }
+    
+    std::string ServerID(serverid);
+    std::ostringstream ocontext;
+    ocontext << " {\"Server ID\": \"" << ServerID << "\"}"; // Set context with server ID
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: If ServerID contains the `"` character, the context will contain invalid JSON.

You probably need to add a function to escape your strings to be JSON-safe.

For example:

    std::string jsonEscape(const std::string &input) {
       std::string output(input);
       std::replace(output.begin(), output.end(), '"', ' '); // replace '"' character with ' ' (space)
       return output;
    }

File: plugins/bf4/bf4.cpp [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    context = ocontext.str();
+
+    std::wostringstream oidentity;
+    oidentity << "{";
+    std::string Host(host);
+    if (Host.find("bot") == std::string::npos) // Set host string as empty if "bot" is found in it
+        if (!Host.empty()) {
+            oidentity << std::endl << "\"Host\": \"" << host << "\""; // If it's not empty, set Host (IP:Port) in identity
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Also needs to be escaped.

File: plugins/bf4/bf4.cpp [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        context.clear(); // Clear context
+        identity.clear(); // Clear identity
+        // Set vectors values to 0.
+        for (int i=0;i<3;i++) {
+            avatar_pos[i] = avatar_front[i] = avatar_top[i] = camera_pos[i] =  camera_front[i] = camera_top[i] = 0.0f;
+        }
+
+        return true; // This tells Mumble to ignore all vectors.
+    }
+
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: In the code below, you are using the C strings `ServerID`, `host` and `team` as-is.

However, if they (for some reason) do not contain a NUL byte, you'll crash Mumble.

Instead, you should probably convert them to std::string before using them.

    std::string sserverid(serverid, sizeof(serverid));
    std::string shost(host, sizeof(host));
    std::string steam(team, sizeof(team));

and use those names in the code below.

**Comment by davidebeatrici** [comment]
Yes, thanks for the corrections :smile:

First correction: 
Yes, I didn't change 1 to 0 when I copied the line, sorry.

Second correction:
I always interpreted boolean as
`Yes -> 1`
and
`No ->  0`
This is why I wrote that in the comment :wink:

File: plugins/bf4/bf4.cpp (hunk 5 above)
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Ah, sorry... I didn't see the constructor JUST BELOW here.

Anyway, the point is: please use the std::string constructor with an explicit size.
Then the NUL bytes do not matter!

**Comment by mkrautz** [comment]
To be clear, what I mean by comment regarding the booleans is that JSON has a native boolean values `true` and `false`. It would make sense to use those, instead of your string types `"Yes"` and `"No"`. Just so we're clear. :-)

**Comment by mkrautz** [comment]
Done with my first round of comments.

I'm thinking it might also make sense to split the code that generates the identity string out into a separate function. (Context seems small enough to me to not have its own function).

File: plugins/bf4/bf4.cpp (hunk 5 above)
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Actually, we can't use my suggestion above, so don't do that. I'll explain...

Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: If we use the `std::string(char *, size_t)` constructor, the string will end up with NUL bytes in it.
Technically that should be OK, but a lot of things break when that's the case (for example, Ice RPC doesn't like having NUL bytes in its strings for some languages)

Instead, you should make sure the strings *are* zero-terminated.

So, you need to do:

    // NUL terminate queried C strings.
    // We do this to ensure the strings from the game
    // are NUL terminated. They should be already, but
    // we can't take any chances.
    serverid[sizeof(serverid)-1] = 0;
    host[sizeof(host)-1] = 0;
    team[sizeof(team)-1] = 0;

**Comment by mkrautz** [comment]
Added new comments about the NUL terminating issues (just so you don't miss 'em).

**Comment by davidebeatrici** [comment]
I didn't see your last comments when I replied, I'm fixing all of the things you mentioned step by step; I will reply again when I will have done everything :wink:

File: plugins/bf4/bf4.cpp [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        if (Team == "US")
+            oidentity << "\"Team\": \"United States\""; // If team value is US, set "United States" as team in identity.
+        else if (Team == "CH")
+            oidentity << "\"Team\": \"China\""; // If team value is CH, set "China" as team in identity.
+        else if (Team == "RU")
+            oidentity << "\"Team\": \"Russia\""; // If team value is RU, set "Russia" as team in identity.
+    }
+
+    if (squad == 0) {
+        oidentity << std::endl << "\"Squad\": \"No\""; // If squad value is 0, set squad state to "No" in identity.
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Maybe `null` instead of `"No"`? I don't know.

File: plugins/bf4/bf4.cpp [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        else if (squad == 23)
+            oidentity << std::endl << "\"Squad\": \"Whiskey\"";
+        else if (squad == 24)
+            oidentity << std::endl << "\"Squad\": \"X-ray\"";
+        else if (squad == 25)
+            oidentity << std::endl << "\"Squad\": \"Yankee\"";
+        else if (squad == 26)
+            oidentity << std::endl << "\"Squad\": \"Zulu\"";
+
+        if (squad_leader)
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Still not boolean `true`/`false`.

**Comment by mkrautz** [comment]
LGTM with my latest comments are addressed.

File: plugins/bf4/bf4.cpp [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+            peekProc((BYTE *) pModule + 0x21C6D40, avatar_pos, 12) && // Avatar Position values (X, Y and Z).
+            peekProc((BYTE *) pModule + 0x21CAF80, camera_pos, 12) && // Camera Position values (X, Y and Z).
+            peekProc((BYTE *) pModule + 0x21CAF60, avatar_top, 12) && // Avatar Top Vector values (X, Y and Z).
+            peekProc((BYTE *) pModule + 0x21CAF70, avatar_front, 12) && // Avatar Front Vector values (X, Y and Z).
+            peekProc((BYTE *) serverid_offset, serverid) && // Server ID (36 characters).
+            peekProc((BYTE *) pModule + 0x21B80C0, host) && // Host value: "IP:Port" when in a server, "bot" when loading and empty when it's hidden.
+            peekProc((BYTE *) pModule + 0x24AFAE5, team) && // Team value: US (United States); RU (Russia); CH (China).
+            peekProc((BYTE *) squad_offset_2 + 0x230, squad) && // Squad value: 0 (not in a squad); 1 (Alpha); 2 (Bravo); 3 (Charlie)... 26 (Zulu).
+            peekProc((BYTE *) squad_offset_2 + 0x234, squad_leader); // Team value: US (United States); RU (Russia); CH (China).
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: The comment here is wrong.

Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Also, don't read squad_leader into a bool. Read it into a BYTE, like `squad`.

File: plugins/bf4/bf4.cpp (hunk 7 above)
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Should be `if (squad_leader == 1) {`.

File: plugins/bf4/bf4.cpp [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+            oidentity << std::endl << "\"Squad\": \"Whiskey\"";
+        else if (squad == 24)
+            oidentity << std::endl << "\"Squad\": \"X-ray\"";
+        else if (squad == 25)
+            oidentity << std::endl << "\"Squad\": \"Yankee\"";
+        else if (squad == 26)
+            oidentity << std::endl << "\"Squad\": \"Zulu\"";
+
+        if (squad_leader)
+            oidentity << std::endl << "\"Squad leader\": \"Yes\""; // If squad leader value is true, set squad leader state to "Yes" as squad leader in identity.
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Should be: `oidentity << std::endl << "\"Squad leader\": true";`

File: plugins/bf4/bf4.cpp [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        else
+            oidentity << std::endl << "\"Squad leader\": \"No\""; // If squad leader value is false, set squad leader state to "No" as squad leader in identity.
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Should be: `oidentity << std::endl << "\"Squad leader\": false,";`

File: plugins/bf4/bf4.cpp [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        oidentity << std::endl;
+        if (Team == "US")
+            oidentity << "\"Team\": \"United States\""; // If team value is US, set "United States" as team in identity.
+        else if (Team == "CH")
+            oidentity << "\"Team\": \"China\""; // If team value is CH, set "China" as team in identity.
+        else if (Team == "RU")
+            oidentity << "\"Team\": \"Russia\""; // If team value is RU, set "Russia" as team in identity.
+    }
+
+    if (squad == 0) {
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: I think this code should be restructured.

Remove this check here and unindent the lines below it.

So the code becomes:

	if (squad == 1) {
		oidentity << std::endl << "\"Squad\": \"Alpha\",";
	} else if (squad == 2)
		oidentity << std::endl << "\"Squad\": \"Bravo\",";
	[...]
	} else {
		oidentity << std::endl << "\"Squad\": null,"
	}

This makes the code handle invalid values (values that are 0, or > 26) consistent, and simplifies the code.

File: plugins/bf4/bf4.cpp [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+    context = ocontext.str();
+
+    std::wostringstream oidentity;
+    oidentity << "{";
+    host[sizeof(host)-1] = 0; // NUL terminate queried C strings. We do this to ensure the strings from the game are NUL terminated. They should be already, but we can't take any chances.
+    std::string Host(host);
+    Host = jsonStringEscape(Host);
+    if (Host.find("bot") == std::string::npos) // Set host string as empty if "bot" is found in it.
+        if (!Host.empty()) {
+            oidentity << std::endl << "\"Host\": \"" << host << "\""; // If it's not empty, set Host (IP:Port) in identity.
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Seems like the JSON needs a comma after the last `"`, so it becomes:

    oidentity << std::endl << "\"Host\": \"" << host << "\",";

File: plugins/bf4/bf4.cpp [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+        }
+
+    team[sizeof(team)-1] = 0; // NUL terminate queried C strings. We do this to ensure the strings from the game are NUL terminated. They should be already, but we can't take any chances.
+    std::string Team(team);
+    Team = jsonStringEscape(Team);
+    if (!Team.empty()) {
+        oidentity << std::endl;
+        if (Team == "US")
+            oidentity << "\"Team\": \"United States\""; // If team value is US, set "United States" as team in identity.
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Needs comma at the end, like was done for "Host".

File: plugins/bf4/bf4.cpp [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+        else if (Team == "CH")
+            oidentity << "\"Team\": \"China\""; // If team value is CH, set "China" as team in identity.
+        else if (Team == "RU")
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Needs comma at the end, like was done for "Host".

File: plugins/bf4/bf4.cpp [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+
+    team[sizeof(team)-1] = 0; // NUL terminate queried C strings. We do this to ensure the strings from the game are NUL terminated. They should be already, but we can't take any chances.
+    std::string Team(team);
+    Team = jsonStringEscape(Team);
+    if (!Team.empty()) {
+        oidentity << std::endl;
+        if (Team == "US")
+            oidentity << "\"Team\": \"United States\""; // If team value is US, set "United States" as team in identity.
+        else if (Team == "CH")
+            oidentity << "\"Team\": \"China\""; // If team value is CH, set "China" as team in identity.
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Needs comma at the end, like was done for "Host".

File: plugins/bf4/bf4.cpp [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+        else if (Team == "RU")
+            oidentity << "\"Team\": \"Russia\""; // If team value is RU, set "Russia" as team in identity.
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Needs comma at the end, like was done for "Host".

File: plugins/bf4/bf4.cpp [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    if (squad == 0) {
+        oidentity << std::endl << "\"Squad\": \"No\""; // If squad value is 0, set squad state to "No" in identity.
+    } else {
+        // If squad value is in a value between 1 and 26, set squad name in identity using NATO Phonetic alphabet.
+        if (squad == 1)
+            oidentity << std::endl << "\"Squad\": \"Alpha\"";
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Needs comma at the end, like was done for "Host".

Same for all the following Squad values.

**Comment by mkrautz** [comment]
OK, done with this round of reviewing. Please address my comments and ping me. :-)

Thanks.

**Comment by mkrautz** [comment]
Thanks!

Landed via a30f1dd445efb1c99a09af86b93f2e10df36f201.

**Comment by davidebeatrici** [comment]
You're welcome :smile:

**Pull request closed without merging**